internal/pkg/composable/providers/filesource/filesource.go

```diff
@@ -0,0 +1,177 @@
+	Type string `config:"type"`
+	Path string `config:"path"`
+}
+
+type contextProvider struct {
+	logger *logger.Logger
+
+	sources map[string]fileSourceConfig
+}
+
+// Run runs the filesource context provider.
+func (c *contextProvider) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create watcher: %w", err)
+	}
+	defer watcher.Close()
+
+	// invert the mapping to map paths to source names
+	inverted := make(map[string][]string, len(c.sources))
+	for sourceName, sourceCfg := range c.sources {
+		sources, ok := inverted[sourceCfg.Path]
+		if !ok {
+			sources = []string{sourceName}
+		} else {
+			sources = append(sources, sourceName)
+		}
+		inverted[sourceCfg.Path] = sources
+	}
+
+	// determine the paths to watch (watch is performed on the directories that contain the file)
+	//
+	// you cannot register the same directory multiple times so this ensures its only registered once
+	paths := make([]string, 0, len(c.sources))
+	for _, cfg := range c.sources {
+		parent := filepath.Dir(cfg.Path)
+		if !slices.Contains(paths, parent) {
+			paths = append(paths, parent)
+		}
+	}
+	for _, path := range paths {
+		err = watcher.Add(path)
+		if err != nil {
+			return fmt.Errorf("failed to watch path %q: %w", path, err)
+		}
+	}
+
+	// read the initial values after the watch has started
+	// this ensures that if the value changed between this code and the loop below
+	// the updated file changes will not be missed
+	current := make(map[string]interface{}, len(c.sources))
+	for path, sources := range inverted {
+		value := c.readContents(path)
+		for _, source := range sources {
+			current[source] = value
+		}
+	}
+	err = comm.Set(current)
+	if err != nil {
+		return fmt.Errorf("failed to set current context: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				// watcher was closed
+				return nil
+			}
+			c.logger.Errorf("file watcher errored: %s", err)
+		case e, ok := <-watcher.Events:
+			if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+				// watcher was closed
+				return nil
+			}
+
+			path := filepath.Clean(e.Name)
+			sources, ok := inverted[path]
+			if !ok {
+				// watching the directory, it can contain files that we are not watching
+				// ignore these events unless we are actively watching this file
+				continue
+			}
+
+			switch {
+			case e.Op&(fsnotify.Create|fsnotify.Write|fsnotify.Remove) != 0:
+				// file was created, updated, or deleted (update the value)
+				value := c.readContents(path)
+				for _, source := range sources {
+					current[source] = value
+				}
+				err = comm.Set(current)
+				if err != nil {
+					return fmt.Errorf("failed to set current context from notify event: %w", err)
+				}
+			}
+		}
+	}
+}
+
+func (c *contextProvider) readContents(path string) string {
+	data, err := os.ReadFile(path)
```

Comment:
Should we guard against very big files? I'm not sure if this is something we should worry about.

Reply:
It is not a bad idea, I think I will put a guard on it. These files are meant to be very small. Maybe I will put a maximum of like 4k in size.